File: src/teams/SingleRobotTeam.scala
package teams
import battlecode.common.{GameActionException, RobotType}
import battlecode.common.RobotType.{ARCHON, GARDENER}
import utils.Current.I
import utils.Building.waitBuild

/**
  * Base class for single robot teams (for testing).
  */
abstract class SingleRobotTeam(robotType: RobotType) extends RobotTeam {

  @throws(classOf[GameActionException])
  override def run(): Unit = {
    robotType match {
      // Archon runs
      case ARCHON =>
        runSingle()
      // Archon spawns Gardener and dies -> Gardener runs
      case GARDENER =>
        I.getType match {
          case ARCHON => archonSpawnSingleGardener()
          case GARDENER => runSingle()
          case _ =>
        }
      // Archon spawns Gardener and dies -> Gardener spawns Robot and dies -> Robot runs
      case _ =>
        I.getType match {
          case ARCHON => archonSpawnSingleGardener()
          case GARDENER => waitBuild(robotType)
          case `robotType` => runSingle()
          case _ =>
        }
    }
  }

  def runSingle(): Unit

  private def archonSpawnSingleGardener(): Unit =
    if (I.readBroadcast(0) == 0) {
      // Set flag to kill next archons
      I.broadcast(0, 1)
      // Construct a gardener and die
      waitBuild(GARDENER)
    }
}

File: src/teams/SingleRobotTeam.java
package teams;

import battlecode.common.GameActionException;
import battlecode.common.RobotType;

import static battlecode.common.RobotType.GARDENER;
import static utils.Building.waitBuild;
import static utils.Current.*;
import static utils.Config.Channel.*;
import static utils.Movement.randomDirection;

/**
 * Base class for single robot scala.teams (for testing).
 */
public abstract class SingleRobotTeam implements RobotTeam {

    private RobotType robotType;

    SingleRobotTeam(RobotType robotType) {
        this.robotType = robotType;
    }

    protected abstract void runSingle();

    @Override
    public void run() throws GameActionException {
        RobotType myType = MyType();
        switch (robotType) {
            case ARCHON:
                runSingle();
                break;
            case GARDENER:
                switch (myType) {
                    case ARCHON:
                        archonSpawnSingle();
                        break;
                    case GARDENER:
                        runSingle();
                }
            default:
                switch (myType) {
                    case ARCHON:
                        archonSpawnSingle();
                        break;
                    case GARDENER:
                        //waitBuild(robotType)
                        break;
                    default:
                        if (myType == robotType)
                            runSingle();
                }
        }
    }

    private void archonSpawnSingle() {
        int channel = ARCHON_SPAWNED.index();
        try {
            if (I.readBroadcast(channel) == 0) {
                // Set flag to kill next archons
                I.broadcast(channel, 1);
                // Construct a gardener and die
                waitBuild(GARDENER, randomDirection(), d -> randomDirection());
            }
        } catch (GameActionException e) {
            System.out.println("Invalid channel: " + channel);
        }
    }

}
